openmls/src/group/mls_group/mod.rs

```diff
@@ -297,17 +295,27 @@ impl MlsGroup {
     // === Load & save ===
 
     /// Loads the state from persisted state.
-    pub fn load<R: Read>(reader: R) -> Result<MlsGroup, Error> {
-        // TODO #245: Remove this once we have a proper serialization format
-        #[allow(deprecated)]
-        let serialized_mls_group: SerializedMlsGroup = serde_json::from_reader(reader)?;
-        Ok(serialized_mls_group.into_mls_group())
+    pub fn load(
+        group_id: &GroupId,
+        backend: &impl OpenMlsCryptoProvider,
+    ) -> Result<MlsGroup, LibraryError> {
+        let group = backend
+            .key_store()
+            .read(group_id.as_slice())
+            .ok_or_else(|| LibraryError::custom("This group ID does not exist in storage"))?;
```

Comment:
After checking, we cannot construct a `KeyStore::Error` as it's just a trait bound on `std::error::Error`. There's a flaw in the error model anyway as keystore errors aren't ingestable by the more general library errors. 

Another solution would be to return simply an `Option<MlsGroup>` (which is the return type of `key_store().read()`) and let the caller produce a more meaningful error if they wish so.

Reply:
Returning an `Option<MlsGroup>` sounds reasonable.

I don't quite understand how the error model is flawed, though. Could you elaborate a little bit?

Reply:
> I don't quite understand how the error model is flawed, though. Could you elaborate a little bit?

The error model in `openmls` in general isn't idiomatic nor ergonomic; You're supposed to be able to convert specific errors (eg Group errors) into a single type that is source of errors. Not a dozen of them.

As an example, this is forcing consumers of the library to do things like that themselves: https://github.com/wireapp/core-crypto/blob/e291c3b26815a70862ca84313b15b083d5f974db/crypto/src/error.rs#L169-L247
As an extra, this also causes code noise in the form of extremely widespread `.map_err(MlsError::from)` calls. (to be accurate, we have 65 instances of this exact call)

In short, it creates brittle boundaries between internal exposed error types and the consumer-facing error APIs.

As a concrete example, there should be an `OpenMLSError` type that can ingest any `KeyStoreError` transparently and this should be returned in all public-facing APIs. This wouldn't even be an issue because right now it's extremely unclear what kind of error I should return. Is this a GroupError? Probably not as it's not related to normal Group operation as per MLS spec? Is it a keystore error? Yes but it's generic and so I cannot build it from outside the implementer. So is it a library error? Yes but something probably fits better.

As you can see the error model in general raises more questions than it answers, contrary to its primary purpose, which is to answer the question "Which specific error did I get and is it my mistake?" when encountering unwanted behavior.

Reply:
Thanks for thorough answer. I see now what you mean, although I don't necessarily agree on the need for a unified error type. We haven't really touched the API since we've brought OpenMLS up to spec, so we're probably going to have a discussion on that at some point in the future anyway. 

Let's go with `Option<MlsGroup>` for now and deal with the general API question when we come to that.

Reply:
Gotcha! :)

I also think `Option<MlsGroup>` is the most sensible choice as of now, considering the current error-management capabilities.